
## Preparing the data.
+ Call somatic SNVs using Strelka and CNAs using TitanCNA.

## Generating data for comparison to other bulk + single cell methods
+ `GenerateBulkInput.R`:
    * Specify config file in `Laks/Config.txt`.
    * Identifies exonic SNVs from `ov2295_clone_snvs.csv`.
    * Retrieves reads from the bulk data at the identified sites.
    * Retrieves major and minor copy numbers from TitanCNA.
    * Outputs file `ov2295_clone_snvs_exon.cvs`.
+ `ExtractReadCounts.py`:
    * Loads the exonic SNVS from `ov2295_clone_snvs_exon.csv`
    * Launches SLURM jobs to retrieve reads from single cells at the sites.
    * Output the reads for each cell in a txt file.
+ `GenerateScInput.R`:
    * Reads in the exonic SNVs and the single cell reads.
    * It will generate three files: `ssm.txt`, `sc.txt`, and `sc_hp.txt`.
        - `ssm.txt` is a copy of `ov2295_clone_exon.csv`.
        - `sc.txt` contains single cell reads to be inputted to our method.
        - `sc_hp.txt` contains estimate of hyperparameters for quantifying the distribution over the single cell reads.
    * Optionally, it also generates: `ssm_trimmed.txt`, `sc_trimmed.txt`, and `sc_hp_trimmed.txt`.
        - These files are the trimmed version of the above based on MIN_CELLS parameter in `Config.txt`. Any exonic SNV that does not have at least MIN_CELLS with variant reads at the site are discarded.

## Comparison to PhyloWGS

- Requires `py27`.
    + I created a conda environment called `py27`.

# Pre-process the data
+ Run `PreparePhyloWGSInput.R`.
+ Run parser in `phylowgs/parser`. Exact steps to are given in [PhyloWGS github repository](https://github.com/morrislab/phylowgs/tree/master/parser). 
+ Enable `py27`.
+ Look up *Normal contamination estimate* in `params.txt` in `optimalClustersolution` of TitanCNA output. We need this value to compute the `cellularity` value as 1 - `Normal contamination estimate`. 
    * For this analysis the value I got for cellularity is 0.139.
```
conda activate py27
cd phylowgs/parser
python parse_cnvs.py -f titan -c 0.139 <path/to/titan/optimalClusterSolution>/<file_name>.segs.txt
# python parse_cnvs.py -f titan -c 0.139 /Users/seonghwanjun/data/cell-line/bulk/Combined/cna/results/titan/hmm/optimalClusterSolution/tumor_sample_1_cluster2.segs.txt
python create_phylowgs_inputs.py --cnvs sample1=cnvs.txt --vcf-type sample1=strelka sample1=<path/to/strelka_vcf>
# python create_phylowgs_inputs.py --cnvs sample1=cnvs.txt --vcf-type sample1=strelka sample1=/Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/somatic.snvs.exon.vcf
mv ssm_data.txt cnv_data.txt params.json /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/
```

# Run inference
+ Point to the SNV and CNA files generated by the parser when running inference.
    * Remove the pickle files before starting a new run.
```
cd phylowgs/
rm *pickle
python evolve.py --params <path/to/params.josn> <path/to/ssm_data.txt> <path/to/cnv_data.txt>
# python evolve.py --params /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/params.json /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/ssm_data.txt /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/cnv_data.txt
# python evolve.py --params /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/trimmed_params.json /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/trimmed_ssm_data.txt /Users/seonghwanjun/data/cell-line/bulk/Combined/PWGS/trimmed_cnv_data.txt
```

# View the results
```
mkdir cell-line-results
cd cell-line-results
python ../write_results.py cell-line ../trees.zip cell-line.summ.json.gz cell-line.muts.json.gz cell-line.mutass.zip
cd ..
mv cell-line-results witness/data/
cd witness
gunzip data/*/*.gz
python index_data.py
python -m SimpleHTTPServer
```

# How to run our method
+ Run `PrepareInput.R`.
    * This file will generate the data based on what was generated for PhyloWGS.
```
cd bulk-sc
./run -c ../Laks/main.config
```

# Evaluate the results
- To evaluate PhyloWGS results, first, unzip the trees: `witness/data/cell-line-results/cell-line.mutass.zip`.
- Copy `phylowgs/mcmc_samples.txt` to `witness/data/cell-line-results/`
- Run `Evaluation.R`.
